refactor: Fix remaining clang-tidy performance-inefficient-vector errors

l0rinc opened the pull request:
PR inspired by https://github.com/bitcoin/bitcoin/pull/29608#issuecomment-2437847307 (and https://github.com/bitcoin/bitcoin/pull/29458, https://github.com/bitcoin/bitcoin/pull/29606, https://github.com/bitcoin/bitcoin/pull/29607, https://github.com/bitcoin/bitcoin/pull/30093).

The `clang-tidy` check can be run via:
```bash
cmake -B build -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_FOR_FUZZING=ON && cmake --build build -j$(nproc)

run-clang-tidy -quiet -p build -j $(nproc) -checks='-*,performance-inefficient-vector-operation' | grep -v 'clang-tidy'
```
which revealed 3 tests and 1 prod warning (+ fuzz and benching, found by hebasto).
Even though the tests aren't performance critical, getting rid of these warnings (for which the checks were already enabled via https://github.com/bitcoin/bitcoin/blob/master/src/.clang-tidy#L18, see below), the fix was quite simple.

<details>
<summary>clang-tidy -list-checks</summary>

```bash
cd src && clang-tidy -list-checks | grep 'vector'
    performance-inefficient-vector-operation
```

</details>

<details>
<summary>Output before the change</summary>

```
src/test/rpc_tests.cpp:434:9: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
  433 |     for (int64_t i = 0; i < 100; i++) {
  434 |         feerates.emplace_back(1 ,1);
      |         ^

src/test/checkqueue_tests.cpp:366:13: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
  365 |         for (size_t i = 0; i < 3; ++i) {
  366 |             tg.emplace_back(
      |             ^

src/test/cuckoocache_tests.cpp:231:9: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
  228 |     for (uint32_t x = 0; x < 3; ++x)
  229 |         /** Each thread is emplaced with x copy-by-value
  230 |         */
  231 |         threads.emplace_back([&, x] {
      |         ^

src/rpc/output_script.cpp:127:17: error: 'push_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
  126 |             for (unsigned int i = 0; i < keys.size(); ++i) {
  127 |                 pubkeys.push_back(HexToPubKey(keys[i].get_str()));
      |                 ^
```

And the fuzz and benchmarks, noticed by hebasto: https://github.com/bitcoin/bitcoin/pull/31305#issuecomment-2483124499

</details>

l0rinc committed:
refactor: Fix remaining clang-tidy performance-inefficient-vector errors

hebasto commented:
Also see https://github.com/bitcoin/bitcoin/pull/31306.

hebasto commented:
What about:
```
/ci_container_base/src/bench/prevector.cpp:91:13: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
   90 |         for (size_t i = 0; i < 260; ++i) {
   91 |             vec.emplace_back();
      |             ^
/ci_container_base/src/bench/prevector.cpp:104:13: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
  102 |         for (size_t i = 0; i < 260; ++i) {
  103 |             // force allocation
  104 |             vec.emplace_back(29, T{});
      |             ^
...
/ci_container_base/src/test/fuzz/txorphan.cpp:44:9: error: 'emplace_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
   43 |     for (uint8_t i = 0; i < 4; i++) {
   44 |         outpoints.emplace_back(Txid::FromUint256(uint256{i}), 0);
      |         ^
```
?

l0rinc committed:
refactor: Preallocate PrevectorFillVector(In)Direct without vector resize

The prevector benchmarks were likely not trying to measure vector resize performance.

l0rinc committed:
refactor: Reserve vectors in fuzz tests

* Since the main LIMITED_WHILE stated `outpoints.size() < 200'000`, I've presized outpoints accordingly.
* `tx_mut.vin` and `tx_mut.vout` weren't caught by the clang-tidy, but addressed them anyway.

l0rinc commented:
Thanks @hebasto, addressed the bench and fuzz ones as well (in separate commits, explaining the decisions)

hebasto reviewed (commented):
Tested 01e54b7d82a053238b62121eeba2ac2c45b88859 on Ubuntu 24.04. The `clang-tidy-19` with the enabled "performance-inefficient-vector-operation" check is happy.

l0rinc commented on src/test/fuzz/txorphan.cpp:
```diff
@@ -38,6 +38,7 @@ FUZZ_TARGET(txorphan, .init = initialize_orphanage)
 
     TxOrphanage orphanage;
     std::vector<COutPoint> outpoints; // Duplicates are tolerated
+    outpoints.reserve(200'000);
```
Are you saying this will affect fuzzing behavior? Does it look under the hood to check how many items to create? I wasn't aware of that and find it odd.
What value do you recommend here instead of the worst case?

theuni reviewed (approved):
ACK 11f3bc229ccd4b20191855fb1df882cfa6145264

l0rinc commented:
Note that while [the benchmarks](https://corecheck.dev/bitcoin/bitcoin/pulls/31305) indicate a speedup, [they're actually the ones that were changed](https://github.com/bitcoin/bitcoin/pull/31305/files#diff-3a60bcd828090e387348302ccc8453c055170d5b5ddb828cd890228191d0fe42R90-R103):
<img width="633" alt="image" src="https://github.com/user-attachments/assets/808d3dac-d91a-4353-ac68-e078124fd7c6">

hebasto reviewed (approved):
ACK 11f3bc229ccd4b20191855fb1df882cfa6145264, tested with clang 19.1.5 + clang-tidy.

achow101 commented:
ACK 11f3bc229ccd4b20191855fb1df882cfa6145264
